contrib/platform/src/com/sun/jna/platform/WindowUtils.java

```diff
@@ -932,7 +1055,203 @@ public boolean outputRange(int x, int y, int w, int h) {
+
+				final BITMAP bmp = new BITMAP();
+				if (iconInfo.hbmColor != null
+						&& iconInfo.hbmColor.getPointer() != Pointer.NULL) {
+					final int nWrittenBytes = GDI32.INSTANCE.GetObject(
+							iconInfo.hbmColor, bmp.size(), bmp.getPointer());
+					bmp.read();
+					if (nWrittenBytes > 0)
+						return new Dimension(bmp.bmWidth.intValue(),
+								bmp.bmHeight.intValue());
+				} else if (iconInfo.hbmMask != null
+						&& iconInfo.hbmMask.getPointer() != Pointer.NULL) {
+					final int nWrittenBytes = GDI32.INSTANCE.GetObject(
+							iconInfo.hbmMask, bmp.size(), bmp.getPointer());
+					bmp.read();
+					if (nWrittenBytes > 0)
+						return new Dimension(bmp.bmWidth.intValue(), bmp.bmHeight.intValue() / 2);
+				}
+			} finally {
+				if (iconInfo.hbmColor != null
+						&& iconInfo.hbmColor.getPointer() != Pointer.NULL)
+					GDI32.INSTANCE.DeleteObject(iconInfo.hbmColor);
+				if (iconInfo.hbmMask != null
+						&& iconInfo.hbmMask.getPointer() != Pointer.NULL)
+					GDI32.INSTANCE.DeleteObject(iconInfo.hbmMask);
+			}
+
+			return new Dimension();
+		}
+
+		@Override
+		public List<DesktopWindow> getAllWindows(
+				final boolean onlyVisibleWindows) {
+			final List<DesktopWindow> result = new LinkedList<DesktopWindow>();
+
+			final WNDENUMPROC lpEnumFunc = new WNDENUMPROC() {
+				@Override
+				public boolean callback(final HWND hwnd, final Pointer arg1) {
+					try {
+						final boolean visible = !onlyVisibleWindows
+								|| User32.INSTANCE.IsWindowVisible(hwnd);
+						if (visible) {
+							final String title = getWindowTitle(hwnd);
+							final String filePath = getProcessFilePath(hwnd);
+							final Rectangle locAndSize = getWindowLocationAndSize(hwnd);
+							result.add(new DesktopWindow(hwnd, title, filePath,
+									locAndSize));
+						}
+					} catch (final Exception e) {
+						e.printStackTrace();
+					}
+
+					return true;
+				}
+			};
+
+			if (!User32.INSTANCE.EnumWindows(lpEnumFunc, null))
+				throw new Win32Exception(Kernel32.INSTANCE.GetLastError());
+
+			return result;
+		}
+
+		@Override
+		public String getWindowTitle(final HWND hwnd) {
+			final int requiredLength = User32.INSTANCE
+					.GetWindowTextLength(hwnd) + 1;
+			final char[] title = new char[requiredLength];
+			final int length = User32.INSTANCE.GetWindowText(hwnd, title,
+					title.length);
+
+			return Native.toString(Arrays.copyOfRange(title, 0, length));
+		}
+
+		@Override
+		public String getProcessFilePath(final HWND hwnd) {
+			final byte[] filePath = new byte[256];
+			final IntByReference pid = new IntByReference();
+			User32.INSTANCE.GetWindowThreadProcessId(hwnd, pid);
+
+			final HANDLE process = Kernel32.INSTANCE.OpenProcess(
+					0x0400 | 0x0010, false, pid.getValue());
+			if (process == null && Kernel32.INSTANCE.GetLastError() != 5)
+				throw new Win32Exception(Kernel32.INSTANCE.GetLastError());
+
+			final int length = Psapi.INSTANCE.GetModuleFileNameExA(process,
+					null, filePath, filePath.length);
```

Comment:
If this is entirely internal, you should prefer using the unicode versions of functions to avoid issues with ASCII conversions.  Call the `-W` version of the function and use a `char[]` buffer instead.